File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/animation_set/AnimationState.java
package at.wirecube.additiveanimations.additive_animator.animation_set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class AnimationState<T> implements AnimationAction<T> {

    public static class Builder<BuilderSubclass extends Builder<?, T>, T> {

        @NonNull
        protected final List<Animation<T>> animations = new ArrayList<>();

        @Nullable
        protected AnimationEndAction<T> endAction;

        @Nullable
        protected AnimationStartAction<T> startAction;

        public Builder() {}

        @NonNull
        public BuilderSubclass addAnimation(@NonNull AnimationAction.Animation<T> animation) {
            animations.add(animation);
            return (BuilderSubclass) this;
        }

        @NonNull
        public BuilderSubclass addAnimations(@NonNull List<AnimationAction.Animation<T>> animations) {
            this.animations.addAll(animations);
            return (BuilderSubclass) this;
        }

        @SafeVarargs
        @NonNull
        public final BuilderSubclass addAnimations(@NonNull AnimationAction.Animation<T>... animations) {
            this.animations.addAll(Arrays.asList(animations));
            return (BuilderSubclass) this;
        }

        @NonNull
        public BuilderSubclass withEndAction(@Nullable AnimationEndAction<T> endAction) {
            this.endAction = endAction;
            return (BuilderSubclass) this;
        }

        @NonNull
        public BuilderSubclass withStartAction(@Nullable AnimationStartAction<T> startAction) {
            this.startAction = startAction;
            return (BuilderSubclass) this;
        }

        public AnimationState<T> build() {
            return new AnimationState<T>() {
                @Override
                public List<Animation<T>> getAnimations() {
                    return animations;
                }

                @Override
                public AnimationEndAction<T> getAnimationEndAction() {
                    return endAction;
                }

                @Override
                public AnimationStartAction<T> getAnimationStartAction() {
                    return startAction;
                }
            };
        }
    }

    public interface AnimationStartAction<T> {
        void onStart(T target);
    }

    public interface AnimationEndAction<T> {
        void onEnd(T target, boolean wasCancelled);
    }

    /**
     * The animations are only allowed to run if the current state of the animated object matches
     * this state.
     */
    public final boolean shouldRun(AnimationState<T> currentState) {
        return currentState == null || currentState == this;
    }

    /**
     * The animationEndListener is only allowed to run if the current state of the animated object matches
     * this state.
     */
    public final boolean shouldRunEndListener(AnimationState<T> currentState) {
        return currentState == null || currentState == this;
    }

    public AnimationEndAction<T> getAnimationEndAction() {
        return null;
    }

    public AnimationStartAction<T> getAnimationStartAction() {
        return null;
    }
}

File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/animation_set/AnimationState.kt
package at.wirecube.additiveanimations.additive_animator.animation_set

abstract class AnimationState<T : Any> : AnimationAction<T> {

    fun interface AnimationStartAction<T> {
        fun onStart(target: T)
    }

    fun interface AnimationEndAction<T> {
        fun onEnd(target: T, wasCancelled: Boolean)
    }

    open class Builder<BuilderSubclass : Builder<BuilderSubclass, T>, T : Any> {

        protected val animations: MutableList<AnimationAction.Animation<T>> = mutableListOf()

        protected var endAction: AnimationEndAction<T>? = null

        protected var startAction: AnimationStartAction<T>? = null

        @Suppress("UNCHECKED_CAST")
        private fun self(): BuilderSubclass = this as BuilderSubclass

        fun addAnimation(animation: AnimationAction.Animation<T>): BuilderSubclass {
            animations.add(animation)
            return self()
        }

        fun addAnimations(animations: List<AnimationAction.Animation<T>>): BuilderSubclass {
            this.animations.addAll(animations)
            return self()
        }

        @SafeVarargs
        fun addAnimations(vararg animations: AnimationAction.Animation<T>): BuilderSubclass {
            this.animations.addAll(animations)
            return self()
        }

        open fun withEndAction(endAction: AnimationEndAction<T>?): BuilderSubclass {
            this.endAction = endAction
            return self()
        }

        open fun withStartAction(startAction: AnimationStartAction<T>?): BuilderSubclass {
            this.startAction = startAction
            return self()
        }

        open fun build(): AnimationState<T> {
            val capturedAnimations = animations.toList()
            val capturedEndAction = endAction
            val capturedStartAction = startAction
            return object : AnimationState<T>() {
                override fun getAnimations(): List<AnimationAction.Animation<T>> = capturedAnimations
                override fun getAnimationEndAction(): AnimationEndAction<T>? = capturedEndAction
                override fun getAnimationStartAction(): AnimationStartAction<T>? = capturedStartAction
            }
        }
    }

    /**
     * The animations are only allowed to run if the current state of the animated object matches
     * this state.
     */
    fun shouldRun(currentState: AnimationState<T>?): Boolean {
        return currentState == null || currentState === this
    }

    /**
     * The animationEndListener is only allowed to run if the current state of the animated object matches
     * this state.
     */
    fun shouldRunEndListener(currentState: AnimationState<T>?): Boolean {
        return currentState == null || currentState === this
    }

    open fun getAnimationEndAction(): AnimationEndAction<T>? = null

    open fun getAnimationStartAction(): AnimationStartAction<T>? = null
}




